ansible_base/rbac/api/serializers.py

```diff
@@ -0,0 +1,232 @@
+from django.conf import settings
+from django.core.exceptions import ObjectDoesNotExist
+from django.db import transaction
+from django.db.utils import IntegrityError
+from django.utils.translation import gettext_lazy as _
+from rest_framework import serializers
+from rest_framework.exceptions import PermissionDenied
+from rest_framework.fields import flatten_choices_dict, to_choices_dict
+from rest_framework.serializers import ValidationError
+
+from ansible_base.lib.abstract_models.common import get_url_for_object
+from ansible_base.lib.serializers.common import CommonModelSerializer
+from ansible_base.lib.utils.validation import ansible_id_validator
+from ansible_base.rbac.models import RoleDefinition, RoleTeamAssignment, RoleUserAssignment
+from ansible_base.rbac.permission_registry import permission_registry  # careful for circular imports
+from ansible_base.rbac.validators import validate_permissions_for_model
+
+
+class ChoiceLikeMixin(serializers.ChoiceField):
+    """
+    This uses a ForeignKey to populate the choices of a choice field.
+    This also manages some string manipulation, right now, adding the local service name.
+    """
+
+    default_error_messages = serializers.PrimaryKeyRelatedField.default_error_messages
+    psuedo_field = None  # define in subclass
+
+    def get_model_for_init(self):
+        "Delay any model references until serializer initialization due to load order concerns"
+        raise NotImplementedError
+
+    def get_dynamic_choices(self):
+        raise NotImplementedError
+
+    def get_dynamic_object(self, data):
+        raise NotImplementedError
+
+    def to_representation(self, value):
+        raise NotImplementedError
+
+    def __init__(self, **kwargs):
+        # Workaround so that the parent class does not resolve the choices right away
+        self.html_cutoff = kwargs.pop('html_cutoff', self.html_cutoff)
+        self.html_cutoff_text = kwargs.pop('html_cutoff_text', self.html_cutoff_text)
+
+        self.allow_blank = kwargs.pop('allow_blank', False)
+        super(serializers.ChoiceField, self).__init__(**kwargs)
```

Comment:
This class is `ChoiceLikeMixin`, not `ChoiceField`.  This probably underscores that we should just use the Python 3 pattern of doing away with the params to `super()`.

Reply:
This is, very unfortunately, intended.

Clients need to be able to get available choices for a RoleDefinition `permissions` which is a multi-select type of field. There are 2 ways you could do this. You could have an `/api/v1/pemissions/` endpoint... and at some point maybe this makes sense, but we're bring-your-own-permission model. The other way is to show choices in OPTIONS.

You can't do a many-related field (like a slug related field) and show choices in OPTIONS. Because DRF doesn't let you.

https://github.com/encode/django-rest-framework/blob/a45432b54de88b28bd2e6db31acfe3dbcfc748dd/rest_framework/metadata.py#L142-L151

This hard duck typing of `isinstance(field, (serializers.RelatedField, serializers.ManyRelatedField))` means that we can't show choices in OPTIONS if its of this (most correct) type. That's why I'm using a `ChoiceField` to start with, but the `ChoiceField` is extremely unwilling to allow the choices to be dynamic.

https://github.com/encode/django-rest-framework/blob/a45432b54de88b28bd2e6db31acfe3dbcfc748dd/rest_framework/fields.py#L1394

The problem isn't performance or anything, the problem there was load order. We really don't want to reference stuff from our registry until after apps are ready, and this `__init__` method very much ran before that. This is one reason I created https://github.com/ansible/django-ansible-base/pull/127, because the settings are available much earlier during import than the registry for this or the other app, dab_resource_registry. So to avoid rendering choices before apps are ready, yes, I jumped over the `ChoiceField.__init__` to its parent and set the other things it did manually.

This is absolutely terrible in every way. I do hope we burn it to the ground, eventually, but this does what we want for right now.

Reply:
Fair enough, I see what you mean.  However, I do prefer DjRF's thinking on this -- the principled thing to do here is have a `permissions` endpoint.

Reply:
> and this __init__ method very much ran before that.

This I'm having a hard time with.  `__init__` methods are not generally run at import time, unless you are (directly or indirectly) instantiating the object at the module level.  Are we doing that somewhere?

Reply:
Sorry for the response delay. DRF fields are initialized at import time. It is rather annoying.

Related - we are planning to move the Permission model into this library as a concrete model. That will eventually allow for unwinding some of the hacks here.